### PR TITLE
Add a RuntimeSpec for contract preparation

### PR DESCRIPTION
**Pull request by nagisa**
I haven't had an opportunity to proof-read this yet, but I believe the description herein to be accurate.

Nits wrt writing style appreciated.

cc @matklad @bowenwang1996 @Ekleog @olonho @jakmeier 

(I ended up rolling the mostly reformatting changes to `FunctionCall` runtime spec into this while at this, lmk if you want this split out)

### CODE REVIEW

**Review by jakmeier — COMMENTED** [review]
Nice write-up! I especially like the examples.

I added some nitpicks for clarity where I stumbled reading. Obviously, these are a matter of taste. So, no offense taken if you ignore them entirely.

File: specs/RuntimeSpec/FunctionCall.md [hunk 1]
```diff
@@ -13,215 +15,246 @@ ActionReceipt {
...
+- count used gas (which is `burnt gas` + gas attached to the new created receipts);
+- measure the increase in account’s storage usage as a result of this call;
+- collect logs produced by the contract;
+- set the return data; and
+- create new receipts through [PromisesAPI](Components/BindingsSpec/PromisesAPI.md).
+
+Note that, as an implementation detail, some of these steps may be executed as part of the
+prerequisite actions such as [`DeployContractAction`](./Actions.md#DeployContractAction). This is
+largely an optimization of the gas fee structure, and may not result in an observable functional
+difference.
```
[the remark this answers is not in the document]
Comment (nagisa) [reply specs/RuntimeSpec/FunctionCall.md:67]: I guess this wording is mostly of relevance to developers of the runtime. Maybe to somebody trying to understand how it is possible to have flat fees for `FunctionCall` be as low as they are.

The intent here is to tell the reader that some of these actions don't necessarily run as part of every `FunctionCall`, that the optimization should not produce any observable difference in behaviour (e.g. the errors are still always reported from `FunctionCall`). And briefly explain the reasoning behind our decision to do this.

I'll see if I can improve the wording.

File: specs/RuntimeSpec/Preparation.md [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+terminated.
+
+Let metered block to refer to a set of wasm instructions. Analysis segments wasm functions
+into metered blocks such that every instruction belongs to exactly one metered block. The
+instructions belonging to a metered block need not immediately follow another instruction within
+the same metered block.
+
+During analysis metered blocks form a stack and the instructions are assigned to the metered block
+on top of the stack. The following terminology will be used throughout this section to refer to
+metered block operations:
```
[the remark this answers is not in the document]
Comment (nagisa) [reply specs/RuntimeSpec/Preparation.md]: I incorporated this wording. Thanks!

File: specs/RuntimeSpec/Preparation.md [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+* The wasm module may contain no more than `1_000_000` global definitions and imports;
+* The wasm module may contain no more than `100_000` data segments;
+* UTF-8 strings comprising the wasm module definition (e.g. export name) may not exceed `100_000`
+  bytes each;
+* Function definitions may not specify more than `50_000` locals;
+* Signatures may not specify more than `1_000` parameters;
+* Signatures may not specify more than `1_000` results;
+* Tables may not specify more than `10_000_000` entries;
+* Module may not specify more than `1` table;
+* Module may not specify more than `1` memory;
```
Comment (Ekleog) [inline specs/RuntimeSpec/Preparation.md:57]: Hmm… are all the numbers above hardcoded or are they genesis parameters? If they're hardcoded, maybe we should at least add a comment alongside the place where they're hardcoded to remind people to edit this place too if the parameter were to change?

Comment (Ekleog) [inline specs/RuntimeSpec/Preparation.md:57]:
```suggestion
* The wasm module may not specify more than `1` table;
* The wasm module may not specify more than `1` memory;
```

File: specs/RuntimeSpec/Preparation.md [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+  nop)
+```
+
+## Operand stack depth instrumentation
+
+The `max_stack_height` genesis parameter imposes a limit on the number of entries the wasm
+operand stack may contain during the contract execution.
+
+The maximum operand stack height required for a function to execute successfully is computed
+statically by emulating operand stack operations each instruction executes. For example `i32.const
```
Comment (Ekleog) [inline specs/RuntimeSpec/Preparation.md:224]:
```suggestion
statically by simulating the operand stack operations that each instruction executes. For example `i32.const
```

**Review by Ekleog — COMMENTED** [review]
For the reformatting changes, I'd just point towards https://code.dblock.org/2021/06/07/to-wrap-or-not-to-wrap-in-markdown.html to advocate not wrapping markdown (unless there are huge paragraphs, which probably should be avoided anyway), in particular because right now the diff is pretty hard to read. So I've reviewed the FunctionCall diff quite quickly and went in-depth only on the preparation page.

Overall LGTM, but I'm not knowledgeable enough to guarantee that all the information here is correct, just that the parts for which I did have previous knowledge match :)

File: specs/RuntimeSpec/Preparation.md (hunk 3 above)
Comment (nagisa) [reply specs/RuntimeSpec/Preparation.md:57]: Some of them come from the version of the wasm specification we use (e.g. multi-memory proposal is not enabled); some others are enforced by [the parser we use](https://github.com/bytecodealliance/wasm-tools/blob/main/crates/wasmparser/src/limits.rs).

File: specs/RuntimeSpec/Preparation.md [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+* UTF-8 strings comprising the wasm module definition (e.g. export name) may not exceed `100_000`
+  bytes each;
+* Function definitions may not specify more than `50_000` locals;
+* Signatures may not specify more than `1_000` parameters;
+* Signatures may not specify more than `1_000` results;
+* Tables may not specify more than `10_000_000` entries;
+* Module may not specify more than `1` table;
+* Module may not specify more than `1` memory;
+
+These additional requirements are imposed at a later stage:
```
[the remark this answers is not in the document]
Comment (nagisa) [reply specs/RuntimeSpec/Preparation.md:59]: The kind of error message that gets put on chain will differ based on the order in which we apply the checks. I'll look into specifying when this later stage is, exactly.

File: specs/RuntimeSpec/FunctionCall.md [hunk 6]
```diff
@@ -13,215 +15,246 @@ ActionReceipt {
...
-- deserializes and validates the `code` Wasm binary (see `prepare::prepare_contract`)
-- injects the gas counting function `gas` which will charge gas on the beginning of the each code block
-- instantiates [Bindings Spec](Components/BindingsSpec/BindingsSpec.md) with binary and calls the `FunctionCall.method_name` exported function
+In order to implement this action, the runtime will:
 
-During execution, VM does the following:
+- load the contract code from the `receiver_id` [account](../DataStructures/Account.md#account)’s
+  storage;
+- parse, validate and instrument the contract code (see [Preparation](./Preparation.md));
+- optionally, compile the contract code to an executable form;
```
[the remark this answers is not in the document]
Comment (nagisa) [reply specs/RuntimeSpec/FunctionCall.md:49]: It is up to the implementation to decide if it wants to run compiled code, or to interpret it. Either way there should be no observable difference in behaviour, so maybe this entire item is irrelevant.

**Comment by nagisa** [comment]
Made changes to address most of your comments, @Ekleog.

> I'd just point towards https://code.dblock.org/2021/06/07/to-wrap-or-not-to-wrap-in-markdown.html to advocate not wrapping markdown

I really, really dislike line wrapping in editors and I can't really work on prose when paragraphs don't fit into the screen, either. I personally also prefer to read raw markdown over the rendered version, and e.g. you can't ask browser to wrap lines when viewing the file [on github](https://github.com/near/NEPs/blob/master/specs/RuntimeSpec/FunctionCall.md?plain=1).

Happy to undo the manual wrapping now that I'm done working on the changes in that file though.

File: specs/RuntimeSpec/Preparation.md [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+In order to implement precise and efficient gas accounting, the contract code is analyzed and
+instrumented with additional operations before the compilation occurs. One such instrumentation
+implements accounting of the gas fees.
+
+Gas fees are accounted for at a granularity of sequences of instructions forming a metered block
+and is consumed before execution of any instruction part of such a sequence. The accounting
+mechanism verifies the remaining gas is sufficient, and subtracts the gas fee from the remaining
+gas budget before continuing execution. In the case where the remaining gas balance is insufficient
+to continue execution, the `GasExceeded` error is raised and execution of the contract is
+terminated.
```
[the remark this answers is not in the document]
Comment (nagisa) [reply specs/RuntimeSpec/Preparation.md:92]: I was going to say something along the lines of

> calling `used_gas` host function multiple times in the same metered block will return the same value

but this function internally charges additional gas as part of its own execution, so it is provably untrue. I have a hard time coming up with any other way to introduce wording around `used_gas` here, but I'll think about it more.

**Review by matklad — APPROVED** [review]
LGTM!

One question I have is whether we have near-side tests for `used_gas` that fixes the block granularity.

**Pull request merged**